core/state/state_object.go

```diff
@@ -369,16 +376,46 @@ func (s *stateObject) updateTrie(db Database) Trie {
 	if s.db.prefetcher != nil {
 		s.db.prefetcher.used(s.data.Root, usedStorage)
 	}
+	if !concurrent {
+		if len(s.pendingStorage) > 0 {
+			s.pendingStorage = make(Storage)
+		}
+	}
+	return tr
+}
+
+func (s *stateObject) updateTrieConcurrent(db Database) {
+	tr := s.getTrie(db)
+	for key, value := range s.pendingStorage {
+		// Skip noop changes, persist actual changes
+		if value == s.originStorage[key] {
+			continue
+		}
+		s.originStorage[key] = value
+
+		var v []byte
+
+		if (value == common.Hash{}) {
+			s.setError(tr.TryDelete(key[:]))
+			s.db.StorageDeleted += 1
```

Comment:
This is shared resource, not safe to update concurrently. Moreover, this is updated in updateTrie already